src/program/config/data_format/data_format.lua

```diff
@@ -0,0 +1,119 @@
+module(..., package.seeall)
+
+local schema = require("lib.yang.schema")
+local yang_data = require("lib.yang.data")
+local usage = require("program.config.data_format.README_inc")
+
+-- Number of spaces a tab should consist of when indenting config.
+local tab_spaces = 4
+
+function run(args)
+   function print_level(level, ...)
+      io.write(string.rep(" ", level * tab_spaces))
+      print(...)
+   end
+   function union_type(union)
+      local rtn
+      for _, t in pairs(union.argument_type.union) do
+         if rtn then
+            rtn = rtn .. " | " .. t.argument_string
+         else
+            rtn = t.argument_string
+         end
+      end
+      return rtn
+   end
+   function collate_options(name, node, keys)
+      local options = {}
+      if keys ~= nil then
+	 for _, k in pairs(keys) do
+	    if name == k then options.key = true end
+	 end
+      end
+      if node.argument_type then
+	 local at = node.argument_type
+	 if at.range then
+	    options.range = at.range.argument_string
+	 end
+      end
+      if node.mandatory then
+	 options.mandatory = true
+      end
+      return options
+   end
+   function display_node(name, node, options, level, parent)
+      function comment()
+	 local comments = {}
+	 if options.mandatory == true then
+	    comments[#comments + 1] = "mandatory"
+	 end
+	 if options.key then
+	    comments[#comments + 1] = "key"
+	 end
+	 if options.range then
+	    comments[#comments + 1] = "between " .. options.range
+	 end
+	 local rtn = nil
+	 for n, c in pairs(comments) do
+	    if n == 1 then
+	       rtn = "// " .. c
+	    else
+	       rtn = rtn  .. " " .. c
+	    end
+	 end
+	 return rtn
+      end
+      function display_leaf(keyword, argument)
+	 if argument == "union" then argument = union_type(node) end
+	 local comments = comment()
+	 local str = keyword .. " " .. argument .. ";"
+	 if comments then
+	    print_level(level, str .. " " .. comments)
+	 else
+	    print_level(level, str)
+	 end
+      end
+      if level == nil then level = 0 end
+      if node.type == "table" or node.type == "struct" then
+	 local keys = {}
+	 if node.keys then
+	    print_level(
+	       level,
+	       "// List: this nested structure is repeated with (a) unique key(s)"
+	    )
+	 end
+	 print_level(level, name.." {")
+         display_data_format(node, level + 1)
+         print_level(level, "}")
+      elseif node.type == "scalar" then
```

Comment:
Here instead of the if/elseif chain, what do you think about writing this in a data-driven way?  Like this:

```lua
local describers = {}
local function describe(keyword, production)
   assert(describers[production.type])(keyword, production)
end
local function describe_sequence(productions)
   local keywords = {} -- sort keys of productions into this list
   for _,keyword in ipairs(keywords) do describe(keyword, productions[keyword]) end
end
function describers.scalar(keyword, production)
   -- here we describe this production...
end
function describers.table(keyword, production) ... end
-- for array and struct as well

--- for top level which is a struct, result is:
describe_sequence(production.members)
```